### ISSUE TITLE
Improving compatibility with Scikit-learn

### ISSUE THREAD

**Issue by daviddiazvico** [tensorflow/tensorflow#32532]
The original issue in keras-team/keras is [this](https://github.com/keras-team/keras/issues/12832
)

Scikit-learn estimators are instanced only with hyper-parameter information. No knowledge on the data (X or y) is needed until the estimator is fitted. On the other hand, KerasRegressor and KerasClassifier receive as argument a build_fn that instantiates and compiles the estimator. At this point, the input and output shapes are needed to define the network's architecture. This is a subtle but important difference between Scikit-learn estimators and KerasClassifier/KerasRegressor.

Also, Scikit-learn's GridSearchCV and RandomizedSearchCV (and probably Scikit-optimize's BayesSearchCV) automatically infer the most appropriate CV folding (stratified or not) from the base_estimator. KerasClassifier and KerasRegressor, however, do not have the attribute needed for this to work.

My proposed solution (with some additional tests) is in [this](https://github.com/daviddiazvico/tensorflow/commit/e33e1c5f88e9832b2578f495ee85bcae7e9a5a4f) commit (I'll open a PR now).

Thank you!

**Comment by fchollet** [comment]
Hello,

Making the sklearn wrapper classes agnostic to input shape seems like a good idea. However the commit you link make many assumptions that wouldn't apply. Since these classes are meant for compatibility with sklearn, we should only support single Numpy arrays as input, since this is what sklearn supports.

**Comment by daviddiazvico** [comment]
Hi,

Thanks for the comment.

I'm not sure I understand it, though. Just to confirm:

I guess you mean that no dict should be allowed (like in [test](https://github.com/tensorflow/tensorflow/pull/32533/commits/e33e1c5f88e9832b2578f495ee85bcae7e9a5a4f#diff-7145b17c14e03227607f6a4617309181R343
)) as input. That is totally true. I included this to provide a better integration with the functional Model class. But I will remove it if you think it is the best option.

However, if you mean that inputs should not be tensors with order>2, I think it may be interesting to keep it as it is.

As far as I know CNNs and maybe even RNNs could work with the hyperparameter search (check [test](https://github.com/tensorflow/tensorflow/pull/32533/commits/e33e1c5f88e9832b2578f495ee85bcae7e9a5a4f#diff-7145b17c14e03227607f6a4617309181R312)), calibrated classifiers ([test](https://github.com/tensorflow/tensorflow/pull/32533/commits/e33e1c5f88e9832b2578f495ee85bcae7e9a5a4f#diff-7145b17c14e03227607f6a4617309181R334)), ensembles ([test](https://github.com/tensorflow/tensorflow/pull/32533/commits/e33e1c5f88e9832b2578f495ee85bcae7e9a5a4f#diff-7145b17c14e03227607f6a4617309181R324)), etc. provided by sklearn, as long as no tranformer is applied before the keras model in a pipeline. If only numpy matrices are allowed as input, we will lose that functionality.

If you confirm me that we are in the first case, I'll do the changes right now.

**Comment by jvishnuvardhan** [comment]
@daviddiazvico Are you raising any PR to update the code and docs? Thanks!

**Comment by daviddiazvico** [comment]
> @daviddiazvico Are you raising any PR to update the code and docs? Thanks!

[This one](https://github.com/daviddiazvico/tensorflow/commit/e33e1c5f88e9832b2578f495ee85bcae7e9a5a4f). It's been open for a while, but I'm afraid I couldn't find where is the documentation to update. The only doc related I could find is just a pointer to [keras docs](https://github.com/daviddiazvico/tensorflow/blob/master/tensorflow/python/keras/README.md).

I opened this same PR in keras ([here](https://github.com/daviddiazvico/keras/commit/053d3b3593650ccbbbc828e524f658fd13c3f6c9)) long time ago, and the changes to the docs are really minimal. If you could please point me out where can I find them in TF I'd fix them.

**Comment by tensorflowbutler** [comment]
It has been 26 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by gowthamkpr** [comment]
@daviddiazvico It seems that the KerasRegressor and Kerasclassifier is deprecated and we are no longer maintaining/bug fixing it. Please forward this issue to https://github.com/adriangb/scikeras.

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for more than a month. Please add additional comments to reopen this issue. Thanks!!
